[BB-5559] Create the plugin to store LTI configurations

tecoholic opened the pull request:
### Description

This PR establishes the `openedx_ltistore` plugin that can store LTI configurations centrally in the database and make it available for all the courses.

### Testing instructions

Since this is the first code commit to the repository, it can be tested by following the instructions in the [README](https://github.com/open-craft/openedx-ltistore/tree/tecoholic/BB-5559-centrally-store-lti-keys#readme) file.

To test the integration with the LTI Consumer XBlock, use the [open-craft:tecoholic/BB-5559-central-stroage-of-lti-keys](https://github.com/open-craft/xblock-lti-consumer/tree/tecoholic/BB-5559-central-stroage-of-lti-keys) branch and install it locally.

### Additional information

1. [OpenCraft ticket](https://tasks.opencraft.com/browse/BB-5559)
1. [LTI Consumer XBlock PR](https://github.com/openedx/xblock-lti-consumer/pull/239)

giovannicimolin commented on README.md:
````diff
@@ -1 +1,64 @@
+3. Add the `openedx_ltistore` to the installed apps. Edit the file `edx-platform/cms/envs/devstack.py`
+    ```diff
+    -INSTALLED_APPS += ['debug_toolbar']
+    +INSTALLED_APPS += ['debug_toolbar', 'openedx_ltistore']
````
This should not be needed if this has the proper hooks in place. Pluggable apps do this automatically.
Checkout the code in the proctoring app:

- https://github.com/openedx/edx-proctoring/blob/6a21189c6e8c2ac421fe0e9755daf9a7a7a56916/setup.py#L93
- https://github.com/openedx/edx-proctoring/blob/6a21189c6e8c2ac421fe0e9755daf9a7a7a56916/edx_proctoring/apps.py#L60

We also have a few client pluggable apps in GitLab, take a look at them as well.

tecoholic replied:
@giovannicimolin Thank you for the pointers. Implemented the plugin mechanism based on the Proctering app.

giovannicimolin commented on openedx_ltistore/serializers.py:
```diff
@@ -0,0 +1,9 @@
+from rest_framework import serializers
```
Remove this dependency to `rest_framework`. Do you think you can use `from django.core import serializers
` do to the job here?

tecoholic replied:
I replaced this with `django.forms.models.model_to_dict`

giovannicimolin commented on lti_store/models.py:
```diff
@@ -0,0 +1,75 @@
+        help_text=_("Client secret provided by the LTI tool provider."),
+    )
+
+    # LTI 1.3 Related variables
+    lti_1p3_internal_private_key = models.TextField(
+        blank=True,
+        help_text=_("Platform's generated Private key. Keep this value secret."),
+    )
+
+    lti_1p3_internal_private_key_id = models.CharField(
+        max_length=255,
+        blank=True,
+        help_text=_("Platform's generated Private key ID"),
+    )
+
+    lti_1p3_internal_public_jwk = models.TextField(
+        blank=True,
+        help_text=_("Platform's generated JWK keyset."),
+    )
+
+    lti_1p3_client_id = models.CharField(
+        max_length=255,
+        default=generate_client_id,
+        help_text=_("Client ID used by LTI tool"),
+    )
```
We only support LTI 1.1 configurations, for now, you can remove these fields until we implement the new version.
```suggestion
```

giovannicimolin commented on README.md:
````diff
@@ -1 +1,66 @@
+
+The development instructions are written with the [Open edX Devstack](https://edx.readthedocs.io/projects/open-edx-devstack/en/latest/index.html) as reference.
+
+1. Clone the repostiory to `<devstack_root>/src/` directory
+   ```
+   cd <devstack_root>/src/
+   git clone git@github.com:open-craft/openedx-ltistore.git
+   ```
+2. Install the plugin inside the container
+   ```
+   cd <devstack_root>/devstack/
+   make studio-shell
+   pip install -e /edx/src/openedx-ltistore
+   ```
+3. Run the migrations:
+   ```
+   make studio-shell
+   python manage.py cms migrate lti_store
+   exit
+   ```
+4. Restart the cms service `make studio-restart`
````
@tecoholic This also needs to be installed in the LMS env, otherwise it'll fail to launch when published

tecoholic commented on lti_store/apps.py:
```diff
@@ -0,0 +1,16 @@
+    default_auto_field = "django.db.models.BigAutoField"
+    name = "lti_store"
+    plugin_app = {
+        "settings_config": {
+            "common": {"relative_path": "settings.common"},
+            "production": {"relative_path": "settings.production"},
+        },
+        "cms.djangoapp": {
+            "common": {"relative_path": "settings.common"},
+            "production": {"relative_path": "settings.production"},
+        },
```
Yes. I added this following the Proctoring app. We can remove this.

tecoholic commented on lti_store/models.py:
```diff
@@ -0,0 +1,75 @@
+    LTI_1P3 = "lti_1p3", _("LTI 1.3")
+
+
+class LTIConfiguration(models.Model):
```
Done. The plugin now uses a slug field and uses the slug for the filter key as well.

tecoholic commented:
@giovannicimolin I have refactored the plugin based on your review feedback and updated the README as well. Give it another pass when you get time. Thank you.

giovannicimolin commented on lti_store/models.py:
```diff
@@ -0,0 +1,39 @@
+
+class LTIVersion(models.TextChoices):
+    LTI_1P1 = "lti_1p1", _("LTI 1.1")
+    LTI_1P3 = "lti_1p3", _("LTI 1.3")
```
@tecoholic Remove the LTI 1.3 version choice from here for now.

tecoholic replied:
Done :)

giovannicimolin reviewed (approved):
:+1:

- [x] I tested this: tested alongside https://github.com/openedx/xblock-lti-consumer/pull/239
- [x] I read through the code
- [ ] I checked for accessibility issues
- [x] Includes documentation

@tecoholic Looks good. Left a minor comment on the PR. 
Good to merge once that is addressed.

giovannicimolin commented:
@tecoholic Remember to squash your commits here before merging.

tecoholic commented:
@giovannicimolin I have updated the PR by removing the version 1.3 choice. I am merging to master.
